### PR TITLE
add advance mode on summon minion safe form

### PR DESCRIPTION
**Pull request by santteegt**
image

### CODE REVIEW

**Comment by scottrepreneur** [comment]
Task linked: [CU-1fr27qy Alt summon flow - connect to existing minion](https://app.clickup.com/t/1fr27qy)

**Comment by skuhlmann** [comment]
@santteegt i think we have some conflicts - maybe because of the minQuorum field update?

also think we might need some info on how to add the safe module? could we link to something or any ideas there?

**Comment by santteegt** [comment]
@skuhlmann I just solved the conflicts.

Regarding adding the minion as a safe module, there's a [FAQ article](https://help.gnosis-safe.io/en/articles/4934427-add-a-module) describing how to do it using the Safe UI, however outdated (it doesn't have the GnosisSafe v1.3.0 Mastercopy address listed). What we can enable is an action button within the Safe section on the Vault page that will craft the Tx to be sent to the Gnosis Safe Tx Service. This action must be performed only by a Safe owner or delegate.

**Comment by skuhlmann** [comment]
> @skuhlmann I just solved the conflicts.
> 
> Regarding adding the minion as a safe module, there's a [FAQ article](https://help.gnosis-safe.io/en/articles/4934427-add-a-module) describing how to do it using the Safe UI, however outdated (it doesn't have the GnosisSafe v1.3.0 Mastercopy address listed). What we can enable is an action button within the Safe section on the Vault page that will craft the Tx to be sent to the Gnosis Safe Tx Service. This action must be performed only by a Safe owner or delegate.

Oh that sounds good.

Do we need to disable anything on the vault page or in proposals before the minion is added as a safe module?

**Comment by santteegt** [comment]
@skuhlmann yeah, actions should be disabled until the minion is added as a module otherwise the execution will throw an exception. Should I try to implement all of these before merging this PR?

**Comment by skuhlmann** [comment]
> @skuhlmann yeah, actions should be disabled until the minion is added as a module otherwise the execution will throw an exception. Should I try to implement all of these before merging this PR?

Yeah I think that would be safest.

checkout those buttons have a memberonly check now. maybe we could add to that?

**Comment by santteegt** [comment]
image

File: src/formBuilder/formBuilder.jsx [hunk 1]
```diff
@@ -241,6 +242,27 @@ const FormBuilder = props => {
     );
   };
 
+  useEffect(() => {
+    // Reset form fields in case these are updated
+    const prevFields = [].concat(...formFields);
+    const prevFieldNames = prevFields.map(f => f.name);
+    const currFields = [].concat(...fields).map(f => f.name);
+    if (!currFields.every(fName => prevFieldNames.includes(fName))) {
```
Comment (skuhlmann) [inline src/formBuilder/formBuilder.jsx:250]: I'm confused on this a bit. What is the instance when this can happen?

Comment (santteegt) [reply src/formBuilder/formBuilder.jsx:250]: this new feature is for cases when you switch forms from default to advanced mode. I was having issues when the No. of fields changes, so field values weren't being assigned properly in the form. Found that I needed to do a deep form reset

File: src/utils/theGraph.js [hunk 2]
```diff
@@ -141,6 +143,21 @@ const completeQueries = {
           supportedChains[args.chainID].network,
           minionAddresses,
         );
+        const vaultData = await Promise.all(
```
Comment (skuhlmann) [inline src/utils/theGraph.js:146]: Maybe this is something we should do in the caching job eventually?

Comment (santteegt) [reply src/utils/theGraph.js:146]: yeah, it should be moved there

File: src/components/safeMinionDetails.jsx [hunk 3]
```diff
@@ -1,13 +1,90 @@
...
 import { FaCopy } from 'react-icons/fa';
+import { TiWarningOutline } from 'react-icons/ti';
+import { utils as Web3Utils } from 'web3';
+import ModuleManager from '@gnosis.pm/safe-contracts/build/artifacts/contracts/base/ModuleManager.sol/ModuleManager.json';
 
 import ContentBox from './ContentBox';
 import TextBox from './TextBox';
+import { ToolTipWrapper } from '../staticElements/wrappers';
+
+import { useInjectedProvider } from '../contexts/InjectedProviderContext';
```
Comment (skuhlmann) [inline src/components/safeMinionDetails.jsx:16]: import order

File: src/components/safeMinionDetails.jsx [hunk 4]
```diff
@@ -1,13 +1,90 @@
...
+  NO_OWNER: {
+    title: 'ACTION REQUIRED',
+    body:
+      'You MUST be a Gnosis Safe signer in order to submit a Tx proposal to add the minion as a module',
+  },
+};
+
+const SafeMinionDetails = ({ vault, safeDetails, handleCopy }) => {
+  const { daochain } = useParams();
+  const [isLoading, setLoading] = useState(false);
```
Comment (skuhlmann) [inline src/components/safeMinionDetails.jsx:39]: component body order

File: src/components/safeMinionDetails.jsx [hunk 5]
```diff
@@ -1,13 +1,90 @@
...
+    const selectedFunction = ModuleManager.abi.find(
+      entry => entry.name === 'enableModule',
+    );
+    const hexData = safeEncodeHexFunction(selectedFunction, [
+      Web3Utils.toChecksumAddress(vault.address),
+    ]);
+    if (!hexData.encodingError) {
+      const checksumSafeAddr = Web3Utils.toChecksumAddress(vault.safeAddress);
+      try {
+        await createGnosisSafeTxProposal({
```
Comment (skuhlmann) [inline src/components/safeMinionDetails.jsx:55]: look into using a tx lego here

File: src/formBuilder/formBuilder.jsx (hunk 1 above)
Comment (skuhlmann) [reply src/formBuilder/formBuilder.jsx:250]: I still see this happening when i toggle the safe minion launch form - i think we need to look into a new method here.

@jordanlesich has done some work on this.

**Pull request merged**